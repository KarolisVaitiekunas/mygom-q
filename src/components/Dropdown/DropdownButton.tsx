import themeGet from '@styled-system/theme-get';
import React from 'react';
import styled from 'styled-components';

export const Root = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 80px;
  width: 50%;
  border: none;
  border-radius: 3px;
  background-color: ${themeGet('colors.primary')};
  cursor: pointer;
`;

interface IDropdownButtonProps {
  handleDropDownVisibility: () => void;
}
const DropdownButton: React.FC<IDropdownButtonProps> = ({ handleDropDownVisibility }) => {
  return <Root onClick={handleDropDownVisibility}>DROPDOWN</Root>;
};

export default DropdownButton;
