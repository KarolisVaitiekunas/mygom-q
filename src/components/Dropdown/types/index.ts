import { DataType } from '../../utils/data';

export type SectionType = Array<{ section: string; elements: DataType[] }>;
