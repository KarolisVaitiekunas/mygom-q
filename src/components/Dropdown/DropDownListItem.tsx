import React from 'react';
import { DataType } from '../utils/data';
import { SectionType } from './types';
import styled from 'styled-components';
import themeGet from '@styled-system/theme-get';

const Section = styled.li`
  padding: 2px;
`;

const SectionHeader = styled.p`
  font-size: ${themeGet('fontSizes.large')}px;
  font-weight: bold;
  text-transform: uppercase;
`;

const SectionText = styled.p`
  font-size: ${themeGet('fontSizes.large')}px;
  cursor: pointer;
  transition: background 0.3s ease-in;
  &:hover {
    background-color: ${themeGet('colors.primary')};
  }
`;

interface IDropDownListItemProps {
  handleElements: (element: DataType) => void;
  section: SectionType[0];
}
const DropDownListItem: React.FC<IDropDownListItemProps> = ({ handleElements, section }) => {
  return (
    <Section>
      <SectionHeader>{section.section}</SectionHeader>
      {section.elements.map((element, index) => {
        return (
          <SectionText key={index} onClick={() => handleElements(element)}>
            {element.name}
          </SectionText>
        );
      })}
    </Section>
  );
};

export default DropDownListItem;
