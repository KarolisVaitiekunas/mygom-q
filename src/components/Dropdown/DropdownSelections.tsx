import React from 'react';
import styled from 'styled-components';
import themeGet from '@styled-system/theme-get';
import { DataType } from '../utils/data';
import { ActiveType } from './types';

export const Root = styled.div<ActiveType>`
  margin-top: 5px;
  display: flex;
  justify-content: space-between;
  width: 100%;
  max-width: 300px;
  overflow-x: auto;
  transition: opacity 0.3s;
  opacity: ${({ show }: { show: boolean }) => (show ? 1 : 0)};
`;

export const SelectedElement = styled.p`
  white-space: nowrap;
  margin: 2px;
  padding: 2px;
  background-color: ${themeGet('colors.primary')};
  color: #fff;
  font-size: ${themeGet('fontSizes.extraSmall')}px;
  font-weight: bold;
  border-radius: 2px;
`;

type DropdownSelectionsPropsType = {
  selectedElements: Array<DataType>;
  show: boolean;
};

const DropdownSelections: React.FC<DropdownSelectionsPropsType> = ({ selectedElements, show }) => {
  return (
    <Root show={show} data-testid='dropdown-selection'>
      {selectedElements.length > 0 ? (
        selectedElements.map((element, index) => (
          <SelectedElement key={index} data-testid='dropdown-selection-item'>
            {element.name}
          </SelectedElement>
        ))
      ) : (
        <h5 style={{ margin: '0 auto 0 auto' }}>No elements selected</h5>
      )}
    </Root>
  );
};

export default DropdownSelections;
