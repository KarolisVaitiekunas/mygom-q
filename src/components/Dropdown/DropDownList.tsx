import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import themeGet from '@styled-system/theme-get';
import DropDownListItem from './DropDownListItem';
import UseFilterData from '../utils/hooks/UseFilterData';
import { DataType } from '../utils/data';
import { ActiveType, SectionType } from './types';

export const Root = styled.div<ActiveType>`
  transition: opacity 0.3s;
  opacity: ${({ show }: ActiveType) => (show ? 1 : 0)};
`;

export const DropdownInput = styled.input`
  margin-top: 10px;
  border: none;
  border-bottom: 2px solid ${themeGet('colors.primary')};

  &:focus {
    outline: none;
  }
`;

export const List = styled.ul`
  max-height: 200px;
  overflow: auto;
`;

type DropDownListPropsType = {
  show: boolean;
  handleElements: (element: DataType) => void;
};
const DropDownList: React.FC<DropDownListPropsType> = ({ show, handleElements }) => {
  const [sections, setSections] = useState<SectionType>([]);
  const { search, list, handleInputSearch, handleFilter } = UseFilterData();

  useEffect(() => {
    // manage sections
    const newSections: SectionType = [];
    list.map((element) => {
      // initial section state will be empty, push so it wouldn't be
      if (newSections.length <= 0) {
        newSections.push({ section: element.type, elements: [element] });
      } else {
        // if does not have section, push one. If has section, push into elements
        const doesNotHaveSection = newSections.every((section) => section.section !== element.type);
        if (doesNotHaveSection) {
          newSections.push({ section: element.type, elements: [element] });
        } else {
          newSections.map((section) => {
            if (section.section === element.type) {
              section.elements.push(element);
            }
          });
        }
      }
    });

    setSections(newSections);
  }, [list]);

  return (
    <Root show={show}>
      <DropdownInput
        placeholder='Search here...'
        value={search}
        onSelect={handleFilter}
        onChange={(e) => handleInputSearch(e.target.value)}
      />
      <List data-testid='dropdown-list'>
        {sections.map((section, index) => (
          <DropDownListItem key={index} handleElements={handleElements} section={section} />
        ))}
      </List>
    </Root>
  );
};

export default DropDownList;
