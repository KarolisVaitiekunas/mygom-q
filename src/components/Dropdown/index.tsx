import { useState } from 'react';
import styled from 'styled-components';
import DropdownButton from './DropdownButton';
import DropDownList from './DropDownList';
import DropdownSelections from './DropdownSelections';
import UseListVisibility from '../utils/hooks/UseListVisibility';
import { DataType } from '../utils/data';

export const Root = styled.div`
  margin-left: auto;
  margin-right: auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 5px;
  position: relative;
  width: 300px;
  max-width: 300px;
`;

const Dropdown: React.FC = () => {
  const [selectedElements, setSelectedElements] = useState<DataType[]>([]);

  const handleElements = (element: DataType) => {
    if (selectedElements.includes(element)) {
      setSelectedElements((state) => state.filter((stateEl) => stateEl !== element));
    } else {
      setSelectedElements((state) => [...state, element]);
    }
  };

  const { show, DropDownListRef, handleDropDownVisibility } = UseListVisibility();

  return (
    <Root ref={DropDownListRef}>
      <DropdownButton handleDropDownVisibility={handleDropDownVisibility} />
      <DropdownSelections show={show} selectedElements={selectedElements} />
      <DropDownList show={show} handleElements={handleElements} />
    </Root>
  );
};

export default Dropdown;
