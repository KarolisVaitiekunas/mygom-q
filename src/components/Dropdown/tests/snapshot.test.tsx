import { render } from '@testing-library/react';
import Dropdown from '../index';

describe('Dropdown ', () => {
  const { container } = render(<Dropdown />);

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });
});
