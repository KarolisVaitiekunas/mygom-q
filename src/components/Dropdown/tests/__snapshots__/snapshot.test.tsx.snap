// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown  should render correctly 1`] = `
<div>
  <div
    class="sc-papXJ TdNoI"
  >
    <button
      class="sc-bczRLJ ciDhpJ"
      data-testid="dropdown-button"
    >
      DROPDOWN
    </button>
    <div
      class="sc-iBkjds jBGvYN"
      data-testid="dropdown-selection"
    >
      <h5
        style="margin: 0px auto 0px auto;"
      >
        No elements selected
      </h5>
    </div>
    <div
      class="sc-eCYdqJ lpuiWt"
    >
      <input
        class="sc-jSMfEi ejJsbW"
        placeholder="Search here..."
        value=""
      />
      <ul
        class="sc-gKXOVf dEecSf"
        data-testid="dropdown-list"
      >
        <li
          class="sc-gsnTZi lcRVku"
        >
          <p
            class="sc-dkzDqf jGBHqn"
          >
            tactile
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            boba u4t
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            mx brown
          </p>
        </li>
        <li
          class="sc-gsnTZi lcRVku"
        >
          <p
            class="sc-dkzDqf jGBHqn"
          >
            linear
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            holy panda
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            mx white
          </p>
        </li>
        <li
          class="sc-gsnTZi lcRVku"
        >
          <p
            class="sc-dkzDqf jGBHqn"
          >
            clicky
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            box blues
          </p>
          <p
            class="sc-hKMtZM gERtIN"
            data-testid="dropdown-list-item"
          >
            box greens
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
